pkg/agent/flowexporter/consumer.go

```diff
@@ -0,0 +1,413 @@
+	klog.V(2).InfoS("Resolved Service address", "address", addr)
+	return addr, dns, nil
+}
+
+func (c *Consumer) createExporter() exporter.Interface {
+	switch c.protocol.Name() {
+	case grpcExporterProtocol:
+		return exporter.NewGRPCExporter(c.nodeName, c.nodeUID, c.obsDomainID)
+	case ipfixExporterProtocol:
+		var collectorProto string
+		if c.protocol.TransportProtocol() == api.FlowExporterTransportTLS {
+			collectorProto = "tcp"
+		} else {
+			collectorProto = string(c.protocol.TransportProtocol())
+		}
+		return exporter.NewIPFIXExporter(collectorProto, c.nodeName, c.obsDomainID, c.v4Enabled, c.v6Enabled)
+	default:
+		klog.V(5).InfoS("invalid protocol for FlowExporterDestination", "exporterProtocol", c.protocol.Name(), "transportProtocol", c.protocol.TransportProtocol())
+		return nil
+	}
+}
+
+func (c *Consumer) Run(stopCh <-chan struct{}) {
+	klog.Info("Consumer started")
+	sub := c.store.Subscribe()
+	defer c.store.Unsubscribe(sub)
+
+	exportTicker := time.NewTicker(c.activeFlowTimeout)
+	defer exportTicker.Stop()
+
+	for {
+		select {
+		case <-stopCh:
+			return
+		case msg := <-sub.C():
+			if msg.Deleted {
+				c.handleDeletedConns(msg.Conns)
+			} else {
+				c.handleUpdatedConns(msg.Conns, msg.L7Events)
+			}
+		case <-exportTicker.C:
+			if !c.connected {
+				ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+				err := c.Connect(ctx)
+				cancel()
+				if err != nil {
+					klog.ErrorS(err, "Error when connecting flow exporter to destination", "name", c.name)
+					c.Reset()
+					// Initializing flow exporter fails, will retry in next cycle.
+					exportTicker.Reset(c.activeFlowTimeout)
+					continue
+				}
+			}
+
+			nextExpireTime, err := c.sendFlowRecords()
+			if err != nil {
+				klog.ErrorS(err, "Error when sending expired flow records")
+				// If there is an error when sending flow records because of
+				// intermittent connectivity, we reset the connection to collector
+				// and retry in the next export cycle to reinitialize the connection
+				// and send flow records.
+				c.Reset()
+				exportTicker.Reset(c.activeFlowTimeout)
+				continue
+			}
+			exportTicker.Reset(nextExpireTime)
+		}
+	}
+}
+
+func (c *Consumer) sendFlowRecords() (time.Duration, error) {
+	currTime := time.Now()
+	var nextExpireTime time.Duration
+	// We export records from denyConnStore first, then conntrackConnStore. We enforce the ordering to handle a
+	// special case: for an inter-node connection with egress drop network policy, both conntrackConnStore and
+	// denyConnStore from the same Node will send out records to Flow Aggregator. If the record from conntrackConnStore
+	// arrives FA first, FA will not be able to capture the deny network policy metadata, and it will keep waiting
+	// for a record from destination Node to finish flow correlation until timeout. Later on we probably should
+	// consider doing a record deduplication between conntrackConnStore and denyConnStore before exporting records.
```

Comment:
This comment was removed from main after I improved the code in order to exclude denied connections from the regular conntrack store: https://github.com/antrea-io/antrea/pull/7361. Please check out that change to see if it has any impact on your PR.
Even without my change, I am not sure that the issue described by the comment would still apply, given the unification of the 2 connection stores. Although maybe your own code currently assumes that the same connection cannot come from both conntrack and packetin (deny). While I believe the assumption to be correct thanks to my change, it's probably better to be a bit careful, and still have a way to merge connection data properly in this case.

Reply:
To reduce the number of changes I've reverted the work on merging the stores and kept the implementation on how things are exported to be as is today.